nightly testrace failed

The nightly testrace run failed on 96302028f139b8e2b38b910b4e45daf16d86ae04. Please review the run [here](https://github.com/cockroachdb/pebble/actions/runs/14053641693).

```
==================
WARNING: DATA RACE
Write at 0x00c001f82070 by goroutine 5062:
  github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).Register()
      /home/runner/work/pebble/pebble/compaction_scheduler.go:301 +0x3d
  github.com/cockroachdb/pebble.Open()
      /home/runner/work/pebble/pebble/open.go:531 +0x63e1
  github.com/cockroachdb/pebble.TestCheckpointManyFiles()
      /home/runner/work/pebble/pebble/checkpoint_test.go:467 +0xa24
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:1743 +0x44

Previous read at 0x00c001f82070 by goroutine 5068:
  github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).tryGrantLockedAndUnlock()
      /home/runner/work/pebble/pebble/compaction_scheduler.go:[38](https://github.com/cockroachdb/pebble/actions/runs/14053641693/job/39348412810#step:4:39)7 +0x2a4
  github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).periodicGranter()
      /home/runner/work/pebble/pebble/compaction_scheduler.go:412 +0x198
  github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).Register.gowrap1()
      /home/runner/work/pebble/pebble/compaction_scheduler.go:303 +0x33

Goroutine 5062 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:1743 +0x825
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:2168 +0x85
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:1690 +0x226
  testing.runTests()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:2166 +0x8be
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:2034 +0xf17
  main.main()
      _testmain.go:721 +0x164

Goroutine 5068 (running) created at:
  github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).Register()
      /home/runner/work/pebble/pebble/compaction_scheduler.go:303 +0xfc
  github.com/cockroachdb/pebble.Open()
      /home/runner/work/pebble/pebble/open.go:531 +0x63e1
  github.com/cockroachdb/pebble.TestCheckpointManyFiles()
      /home/runner/work/pebble/pebble/checkpoint_test.go:4[40](https://github.com/cockroachdb/pebble/actions/runs/14053641693/job/39348412810#step:4:41) +0x575
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:17[43](https://github.com/cockroachdb/pebble/actions/runs/14053641693/job/39348412810#step:4:44) +0x44
==================
--- FAIL: TestCheckpointManyFiles (7.91s)
```

--- a/compaction_scheduler.go
+++ b/compaction_scheduler.go
@@ -237,7 +237,8 @@ func (c *pickedCompactionCache) add(pc *pickedCompaction) {
 // ConcurrencyLimitScheduler is the default scheduler used by Pebble. It
 // simply uses the concurrency limit retrieved from
 // DBForCompaction.GetAllowedWithoutPermission to decide the number of
-// compactions to schedule.
+// compactions to schedule. ConcurrencyLimitScheduler must have its Register
+// method called at most once -- i.e., it cannot be reused across DBs.
 //
 // Since the GetAllowedWithoutPermission value changes over time, the
 // scheduler needs to be quite current in its sampling, especially if the

--- a/open.go
+++ b/open.go
@@ -81,6 +81,9 @@ func Open(dirname string, opts *Options) (db *DB, err error) {
 	// Make a copy of the options so that we don't mutate the passed in options.
 	opts = opts.Clone()
 	opts.EnsureDefaults()
+	if opts.Experimental.CompactionScheduler == nil {
+		opts.Experimental.CompactionScheduler = newConcurrencyLimitScheduler(defaultTimeSource{})
+	}
 	if err := opts.Validate(); err != nil {
 		return nil, err
 	}

--- a/open_test.go
+++ b/open_test.go
@@ -133,8 +133,6 @@ func TestErrorIfNotPristine(t *testing.T) {
 		ErrorIfNotPristine: true,
 	})
 	defer ensureFilesClosed(t, opts)()
-	// Ensures each DB will create its own CompactionScheduler in Open.
-	opts.Experimental.CompactionScheduler = nil
 
 	d0, err := Open("", opts)
 	require.NoError(t, err)

--- a/options.go
+++ b/options.go
@@ -1272,9 +1272,6 @@ func (o *Options) EnsureDefaults() {
 	if o.Experimental.MultiLevelCompactionHeuristic == nil {
 		o.Experimental.MultiLevelCompactionHeuristic = WriteAmpHeuristic{}
 	}
-	if o.Experimental.CompactionScheduler == nil {
-		o.Experimental.CompactionScheduler = newConcurrencyLimitScheduler(defaultTimeSource{})
-	}
 
 	o.initMaps()
 }

